keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,216 @@
+   
+2. Compare whether the `PodGroup` of the current pod is `LastPodGroup`, and if so, keep it unchanged. If not, it indicates that the new PodGroup scheduling cycle has been entered. Update LastPodGroup. And if there is a pod belongs to the last podgroup is waiting and if no chance to meet minAvailable, reject it directly here. The purpose is 1. To avoid invalid waiting, 2. To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.
+
+```go
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+    …
+ 
+     // Count the numbers of all pods belongs to the same PodGroup excludes terminating ones.
+    total := cs.calculateTotalPods(podGroupName)
+    if total < minAvailable {
+        return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+    }
+    ...
+    if cs.Last != "" && cs.Last != podGroupName {
+        allocated := cs.calculateAllocatedPods(cs.LastPodGroup)
+        pgInfo, ok := cs.PodGroupInfos[cs.LastPodGroup]
+        if ok {
+        	pgInfo.LastFailureTimestamp = time.Now()
+        	if allocated < pgInfo.minAvailable {
+            	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+                	if p.GetPod().Labels[PodGroupName] == cs.LastPodGroup {
+                    	p.Reject(cs.Name())
+                	}
+            	})
+        	}
+        }
+    }
+    cs.LastPodGroup = podGroupName
+    return nil
+}
+```
+
+#### Permit
+1. Get the number of Running pods that belong to the same PodGroup
+2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), apprve the waiting pods that  belongs to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout(eg: 10s)
+ 
+```go
+func (cs Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+    ...
+    running := cs.calculateRunningPods(podGroupName)
+    waiting := cs.calculateWaitingPods(podGroupName)
+    if running+waiting+1 < minAvailable {
+        return framework.NewStatus(framework.Wait, ""), 10 * time.Second
+    }
+    ...
+    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+        if p.GetPod().Labels[PodGroupName] == podGroupName {
+        	p.Allow(cs.Name())
+        }
+    })
+   
+    delete(cs.PodGroupInfos, podGroupName)
+    return framework.NewStatus(framework.Success, ""), 0
+}
+```
+ 
+
+#### UnReserve
+After a pod which belongs to a PodGroup times out in the permit phase.  We will release the pods belongs to the same PodGroup to avoid long-term invalid occupy of resources.
```

Comment:
```suggestion
After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.
```